includes/express-checkout/class-wc-payments-express-checkout-button-display-handler.php

```diff
@@ -103,6 +124,11 @@ public function init() {
 			add_action( 'woocommerce_proceed_to_checkout', [ $this, 'display_express_checkout_buttons' ], 21 );
 			add_action( 'woocommerce_checkout_before_customer_details', [ $this, 'display_express_checkout_buttons' ], 1 );
 			add_action( 'woocommerce_pay_order_before_payment', [ $this, 'display_express_checkout_buttons' ], 1 );
+
+			add_action( 'woocommerce_after_add_to_cart_form', [ $this, 'add_order_attribution_inputs' ], 1 );
+			add_action( 'woocommerce_after_cart', [ $this, 'add_order_attribution_inputs' ], 1 );
+			add_action( 'woocommerce_checkout_before_customer_details', [ $this, 'add_order_attribution_inputs' ], 1 );
+			add_action( 'woocommerce_pay_order_before_payment', [ $this, 'add_order_attribution_inputs' ], 1 );
```

Comment:
It appears that we have added hooks twice in this `init` function for `woocommerce_after_add_to_cart_form`, `woocommerce_checkout_before_customer_details`, and `woocommerce_pay_order_before_payment`. Can we combine any of these into a single action, instead of attaching separate hooks? 

If the former is not preferred or possible, it appears that `add_order_attribution_inputs()` is dependent on `display_express_checkout_buttons()` having been fired first, otherwise `should_show_woopay`, `should_show_payment_request`, and `should_show_express_checkout_button` will not yet have been defined. Should we change the priority on the `add_order_attribution_inputs` actions to ensure that this is fired after `display_express_checkout_buttons` or are we able to set those three variables inside this class's constructor? 🤔

Reply:
I was originally thinking the `woocommerce_after_cart` action fired before the `woocommerce_proceeed_to_checkout` action but it doesn't. So there's no need for duplicating the hooks since `woocommerce_after_cart` can be called within `display_express_checkout_buttons`. Much cleaner solution in [a022d92](https://github.com/Automattic/woocommerce-payments/pull/9038/commits/a022d92a80d764ad12021ec057e7c18292fd930e).